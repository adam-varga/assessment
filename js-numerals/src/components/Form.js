import React from "react";
import {
  Container,
  Title,
  InputContainer,
  SubmitButton,
  ConvertedLarge,
  ConvertedSmall,
  ErrorMessage,
  ErrorMessageLarge,
} from "./Form.styles";
import NumberInput from "./NumberInput";
import convertNumber from "../number-conversion/convertNumber";

export default class Form extends React.Component {
  state = {
    inputValue: "",
    converted: null,
    error: null,
  };

  submit = () => {
    try {
      const converted = convertNumber(parseInt(this.state.inputValue));

      this.setState({
        converted,
      });
    } catch (error) {
      this.setState({ converted: null, error: error.message });
    }
  };

  onInputChange = (event) => {
    this.setState({ inputValue: event.target.value, error: null });
  };

  render() {
    return (
      <React.Fragment>
        <ConvertedLarge>{this.state.converted}</ConvertedLarge>
        <ErrorMessageLarge>{this.state.error}</ErrorMessageLarge>
        <Container>
          <Title>
            If you're wondering how to say a number using English words, just
            enter it in the box and hit CONVERT!
          </Title>
          <InputContainer>
            <NumberInput
              value={this.state.inputValue}
              onChange={this.onInputChange}
            ></NumberInput>
            <SubmitButton onClick={this.submit} data-testid="submit-button">
              Convert
            </SubmitButton>
          </InputContainer>
          {this.state.error && <ErrorMessage data-testid='error'>{this.state.error}</ErrorMessage>}
          <ConvertedSmall data-testid='converted'>{this.state.converted}</ConvertedSmall>
        </Container>
      </React.Fragment>
    );
  }
}
