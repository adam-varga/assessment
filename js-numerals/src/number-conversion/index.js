import { baseNumbers, namedPowersOfTen } from "./dictionary";

function underHundred(N) {
  if (N <= 20) {
    return baseNumbers[N];
  }

  let result = baseNumbers[Math.floor(N / 10) * 10];

  if (N % 10) {
    result += `-${baseNumbers[N % 10]}`;
  }

  return result;
}

function underThousand(N) {
  let result = "";

  if (N >= 100) {
    result += `${baseNumbers[Math.floor(N / 100)]} ${namedPowersOfTen[2]}`;
  }

  if (N % 100) {
    result += `${result.length ? " and " : ""}${underHundred(N % 100)}`;
  }

  return result;
}

function underTenThousand(N) {
  let result = "";

  if (N >= 100) {
    if (Math.floor(N / 100) % 10 === 0) {
      result += `${baseNumbers[Math.floor(N / 100) / 10]} ${
        namedPowersOfTen[3]
      }`;
    } else {
      result += `${baseNumbers[Math.floor(N / 100)]} ${namedPowersOfTen[2]}`;
    }
  }

  if (N % 100) {
    result += `${result.length ? " and " : ""}${underHundred(N % 100)}`;
  }

  return result;
}

export default function convertNumber(N) {
  if (N >= Math.pow(10, 21)) {
    throw new Error('Too large number: N < 10^21');
  }

  if (!N) {
    return baseNumbers[0];
  }

  if (N < 10000) {
    return underTenThousand(N)
  } else {
    let result = '';
    let remainder = N;

    const highestNamedPowerOfTen = Math.log10(N) - Math.log10(N) % 3;


    for (let i = highestNamedPowerOfTen; i > 0; i -= 3) {
      const M = Math.floor(remainder / Math.pow(10, i))

      if (M) {
        result += `${result.length ? ' ' : ''}${underThousand(M)} ${namedPowersOfTen[i]}`;

        remainder -= M * Math.pow(10, i);
      }
    }

    if (remainder) {
        result += `${remainder < 100 ? ' and ' : ' '}${underTenThousand(remainder)}`;
    }

    return result;
  }
}

window.convertNumber = convertNumber;
