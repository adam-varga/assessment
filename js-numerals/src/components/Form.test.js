import React from "react";
import { render, cleanup, fireEvent } from "@testing-library/react";
import Form from "./Form";

import * as convertNumber from "../number-conversion/convertNumber";

afterEach(cleanup);

it("renders correctly", () => {
  const { asFragment } = render(<Form />);

  expect(asFragment()).toMatchSnapshot();
});

it("only accepts numbers in number input", () => {
  const { getByTestId } = render(<Form />);

  const numberInput = getByTestId("number-input");

  fireEvent.change(numberInput, { target: { value: "sda" } });
  expect(numberInput.value).toBe("");

  fireEvent.change(numberInput, { target: { value: "4021" } });
  expect(numberInput.value).toBe("4021");

  fireEvent.change(numberInput, { target: { value: "4021s" } });
  expect(numberInput.value).toBe("4021");

  fireEvent.change(numberInput, { target: { value: "40212" } });
  expect(numberInput.value).toBe("40212");
});

it("calls convertNumber with number input on submit button click", () => {
  const { getByTestId } = render(<Form />);

  convertNumber.default = jest.fn();

  const numberInput = getByTestId("number-input");
  const submitButton = getByTestId("submit-button");

  fireEvent.change(numberInput, { target: { value: "4021" } });
  fireEvent(
    submitButton,
    new MouseEvent("click", {
      bubbles: true,
      cancelable: true,
    })
  );

  expect(convertNumber.default).toHaveBeenCalledWith(4021);
});

it("displays converted number text on submit button click", () => {
  const { getByTestId } = render(<Form />);

  convertNumber.default = jest.fn((N) => `${N}_converted`);

  const numberInput = getByTestId("number-input");
  const submitButton = getByTestId("submit-button");
  const convertedText = getByTestId("converted");

  expect(convertedText.textContent).toBe("");

  fireEvent.change(numberInput, { target: { value: "1999" } });

  fireEvent(
    submitButton,
    new MouseEvent("click", {
      bubbles: true,
      cancelable: true,
    })
  );

  expect(convertedText.textContent).toBe("1999_converted");
});

describe("error message", () => {
  let numberInput;
  let submitButton;
  let convertedText;
  let getByTestId;
  let queryByTestId;

  beforeEach(() => {
    const renderResult = render(<Form />);

    getByTestId = renderResult.getByTestId;
    queryByTestId = renderResult.queryByTestId;

    numberInput = getByTestId("number-input");
    submitButton = getByTestId("submit-button");
    convertedText = getByTestId("converted");

    convertNumber.default = jest.fn(() => {
      throw new Error("Test error");
    });

    fireEvent.change(numberInput, { target: { value: "1999" } });

    fireEvent(
      submitButton,
      new MouseEvent("click", {
        bubbles: true,
        cancelable: true,
      })
    );
  });

  it("displays error message if convertNumber throws", () => {
    const errorMessage = getByTestId("error");

    expect(convertedText.textContent).toBe("");
    expect(errorMessage.textContent).toBe("Test error");
  });

  it("hides error if input value is changed after error", () => {
    fireEvent.change(numberInput, { target: { value: "19991" } });

    expect(queryByTestId("error")).toBeFalsy();
  });
});
