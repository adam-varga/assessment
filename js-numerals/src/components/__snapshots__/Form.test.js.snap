// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<DocumentFragment>
  <h1
    class="sc-AxheI cjWyKD"
  />
  <h1
    class="sc-AxmLO ilIuBR"
  />
  <div
    class="sc-AxjAm cJZJJt"
  >
    <h1
      class="sc-AxirZ fvBeIn"
    >
      If you're wondering how to say a number using English words, just enter it in the box and hit CONVERT!
    </h1>
    <div
      class="sc-AxiKw hCRdrJ"
    >
      <input
        class="sc-AxhCb dmlFAv"
        data-testid="number-input"
        value=""
      />
      <button
        class="sc-AxhUy igtKpQ"
        data-testid="submit-button"
      >
        Convert
      </button>
    </div>
    <span
      class="sc-AxgMl eswhUR"
      data-testid="converted"
    />
  </div>
</DocumentFragment>
`;
